File: Tests/RedlineCoreTests/ProviderOverviewTests.swift
// Which state a provider card resolves to, and what it says when a figure is missing.
import XCTest
@testable import RedlineCore

final class ProviderCardStateTests: XCTestCase {
    private func usage(io: Int, cost: Double = 1) -> ProviderUsage {
        var u = ProviderUsage()
        u.io = io
        u.cost = cost
        return u
    }

    func testAMissingToolIsNotInstalledWhateverElseIsTrue() {
        let card = ProviderOverview.card(provider: "Codex", installed: false, read: true,
                                        usage: usage(io: 5000))
        XCTAssertEqual(card.connection, .notInstalled)
        XCTAssertFalse(card.connection.hasFigures,
                       "a tool that is not on this Mac has no figures to draw")
    }

    func testAnInstalledToolSwitchedOffReadsAsNotRead() {
        let card = ProviderOverview.card(provider: "Codex", installed: true, read: false,
                                        usage: usage(io: 5000))
        XCTAssertEqual(card.connection, .notRead)
    }

    func testUnreachableOutranksIdleSoAStoppedServerIsNeverReportedAsQuiet() {
        // The bug this guards: a stopped Ollama has no usage, and "no usage in this range"
        // reads as a quiet day rather than as a server that is not running.
        let card = ProviderOverview.card(provider: "Ollama", installed: true, read: true,
                                        reachable: false, usage: nil)
        XCTAssertEqual(card.connection, .unreachable)
    }

    func testNoUsageAndNoWindowIsIdle() {
        let card = ProviderOverview.card(provider: "Codex", installed: true, read: true,
                                        usage: nil)
        XCTAssertEqual(card.connection, .idle)
    }

    func testAWindowAloneMakesACardActiveEvenWithNoTokens() {
        // A claude.ai user has rate limits and no transcripts, and the card still has news
        let window = LimitWindow(provider: "Claude", key: "five_hour", utilization: 40,
                                 resetsAt: Date().addingTimeInterval(3600))
        let card = ProviderOverview.card(provider: "Claude", installed: true, read: true,
                                        usage: nil, windows: [window])
        XCTAssertEqual(card.connection, .active)
        XCTAssertEqual(card.utilization, 40)
    }

    func testTheWorstWindowIsTheOneShown() {
        let session = LimitWindow(provider: "Codex", key: "five_hour", utilization: 22,
                                 resetsAt: nil)
        let week = LimitWindow(provider: "Codex", key: "seven_day", utilization: 71,
                               resetsAt: nil)
        let card = ProviderOverview.card(provider: "Codex", installed: true, read: true,
                                        usage: usage(io: 10), windows: [session, week])
        XCTAssertEqual(card.worstWindow?.key, "seven_day")
        XCTAssertEqual(card.remainingPercent, 29)
    }

    func testWindowsBelongingToOtherProvidersAreIgnored() {
        let mine = LimitWindow(provider: "Codex", key: "five_hour", utilization: 10,
                               resetsAt: nil)
        let theirs = LimitWindow(provider: "Claude", key: "five_hour", utilization: 99,
                                 resetsAt: nil)
        let card = ProviderOverview.card(provider: "Codex", installed: true, read: true,
                                        usage: usage(io: 10), windows: [mine, theirs])
        XCTAssertEqual(card.utilization, 10, "another provider's window leaked into this card")
    }

    func testUninformativeWindowsAreDropped() {
        let junk = LimitWindow(provider: "Claude", key: "nimbus_quill", utilization: 0,
                               resetsAt: nil)
        let card = ProviderOverview.card(provider: "Claude", installed: true, read: true,
                                        usage: nil, windows: [junk])
        XCTAssertNil(card.worstWindow)
        XCTAssertEqual(card.connection, .idle)
    }

    func testAStaleReadingCarriesNoPaceBecauseARateNeedsACurrentNumber() {
        let resets = Date().addingTimeInterval(3600)
        let window = LimitWindow(provider: "Claude", key: "five_hour", utilization: 50,
                                 resetsAt: resets)
        let pace = PaceEstimator.pace(for: window)
        XCTAssertNotNil(pace, "fixture needs a pace for the test to mean anything")
        let card = ProviderOverview.card(provider: "Claude", installed: true, read: true,
                                        usage: nil, windows: [window],
                                        paces: [pace].compactMap { $0 },
                                        asOf: Date().addingTimeInterval(-3600))
        XCTAssertTrue(card.isStale)
        XCTAssertNil(card.pace)
    }

    func testAFreshReadingKeepsItsPace() {
        let resets = Date().addingTimeInterval(3600)
        let window = LimitWindow(provider: "Claude", key: "five_hour", utilization: 50,
                                 resetsAt: resets)
        guard let pace = PaceEstimator.pace(for: window) else {
            return XCTFail("fixture needs a pace")
        }
        let card = ProviderOverview.card(provider: "Claude", installed: true, read: true,
                                        usage: nil, windows: [window], paces: [pace],
                                        asOf: Date())
        XCTAssertFalse(card.isStale)
        XCTAssertNotNil(card.pace)
    }

    func testStaleOutranksTheStatusItWouldOtherwiseDraw() {
        let window = LimitWindow(provider: "Claude", key: "five_hour", utilization: 96,
                                 resetsAt: nil)
        let card = ProviderOverview.card(provider: "Claude", installed: true, read: true,
                                        usage: nil, windows: [window],
                                        asOf: Date().addingTimeInterval(-4000))
        XCTAssertEqual(card.status(approaching: 60, atLimit: 85).kind, .stale,
                       "an old reading must never be drawn as a live one")
    }

    func testALocalProviderExplainsWhyItHasNoLimitRatherThanShowingNothing() {
        let card = ProviderOverview.card(provider: "Ollama", installed: true, read: true,
                                        reachable: true, usage: usage(io: 900))
        XCTAssertEqual(card.connection, .active)
        XCTAssertNil(card.utilization)
        XCTAssertEqual(card.limitNote,
                       "Runs on this Mac, so there is no rate limit to report")
    }

    func testAHostedProviderWithNoWindowSaysSoAndPrefersTheCallersReason() {
        let plain = ProviderOverview.card(provider: "Claude", installed: true, read: true,
                                         usage: usage(io: 900))
        XCTAssertEqual(plain.limitNote, "No limit window reported yet")

        let given = ProviderOverview.card(provider: "Claude", installed: true, read: true,
                                          usage: usage(io: 900),
                                          limitsNote: "rate limited, retrying")
        XCTAssertEqual(given.limitNote, "rate limited, retrying")
    }

    func testACardWithAWindowCarriesNoLimitNote() {
        let window = LimitWindow(provider: "Codex", key: "five_hour", utilization: 12,
                                 resetsAt: nil)
        let card = ProviderOverview.card(provider: "Codex", installed: true, read: true,
                                        usage: nil, windows: [window],
                                        limitsNote: "should not be shown")
        XCTAssertNil(card.limitNote)
    }

    func testAnAbsentToolSaysNothingAboutLimitsBecauseTheCardAlreadySaysWhy() {
        let card = ProviderOverview.card(provider: "Codex", installed: false, read: true,
                                        usage: nil, limitsNote: "irrelevant")
        XCTAssertNil(card.limitNote)
    }

    func testEveryConnectionStateSaysSomethingAndHasATone() {
        for state in [ProviderConnection.notInstalled, .notRead, .unreachable, .idle, .active] {
            XCTAssertFalse(state.phrase.isEmpty)
            _ = state.tone
        }
    }
}

final class OverviewWarningTests: XCTestCase {
    private func window(_ provider: String, _ key: String, _ pct: Double,
                        resets: Date? = nil) -> LimitWindow {
        LimitWindow(provider: provider, key: key, utilization: pct, resetsAt: resets)
    }

    func testAHealthyWindowRaisesNothing() {
        let warnings = ProviderOverview.warnings(
            windows: [window("Claude", "five_hour", 12)], paces: [],
            approaching: 60, atLimit: 85)
        XCTAssertTrue(warnings.isEmpty, "a banner that is always lit says nothing")
    }

    func testCrossingTheConfiguredThresholdRaisesOne() {
        let warnings = ProviderOverview.warnings(
            windows: [window("Claude", "five_hour", 62)], paces: [],
            approaching: 60, atLimit: 85)
        XCTAssertEqual(warnings.count, 1)
        XCTAssertEqual(warnings.first?.kind, .approaching)
        XCTAssertTrue(warnings.first?.text.contains("62%") ?? false)
    }

    func testTheThresholdsAreTheConfiguredOnesNotHardcodedSixtyAndEightyFive() {
        let warnings = ProviderOverview.warnings(
            windows: [window("Claude", "five_hour", 45)], paces: [],
            approaching: 40, atLimit: 70)
        XCTAssertEqual(warnings.first?.kind, .approaching)

        let atLimit = ProviderOverview.warnings(
            windows: [window("Claude", "five_hour", 72)], paces: [],
            approaching: 40, atLimit: 70)
        XCTAssertEqual(atLimit.first?.kind, .atLimit)
    }

    func testWorstNewsIsReadFirst() {
        let windows = [window("Claude", "five_hour", 65), window("Codex", "seven_day", 91)]
        let warnings = ProviderOverview.warnings(windows: windows, paces: [],
                                                 approaching: 60, atLimit: 85)
        XCTAssertEqual(warnings.first?.kind, .atLimit)
        XCTAssertEqual(warnings.first?.provider, "Codex")
    }

    func testAStaleProviderRaisesNothingBecauseItIsNotEvidenceAboutNow() {
        let warnings = ProviderOverview.warnings(
            windows: [window("Claude", "five_hour", 99)], paces: [],
            approaching: 60, atLimit: 85, staleProviders: ["Claude"])
        XCTAssertTrue(warnings.isEmpty)
    }

    func testAWindowRunningOutBeforeItResetsIsWorthSayingEvenWhileHealthy() {
        let resets = Date().addingTimeInterval(4 * 3600)
        let w = window("Codex", "five_hour", 20, resets: resets)
        // Burning fast enough to hit the cap well before the reset
        let pace = Pace(provider: "Codex", key: "five_hour", utilization: 20,
                        resetsAt: resets, ratePerHour: 80, basis: .windowAverage,
                        elapsedFraction: 0.2,
                        exhaustsAt: Date().addingTimeInterval(3600))
        let warnings = ProviderOverview.warnings(windows: [w], paces: [pace],
                                                 approaching: 60, atLimit: 85)
        XCTAssertEqual(warnings.count, 1)
        XCTAssertEqual(warnings.first?.kind, .runsOutEarly)
        XCTAssertTrue(warnings.first?.text.contains("before it resets") ?? false)
    }

    func testUninformativeWindowsRaiseNothing() {
        let junk = window("Claude", "nimbus_quill", 0)
        XCTAssertTrue(ProviderOverview.warnings(windows: [junk], paces: [],
                                                approaching: 60, atLimit: 85).isEmpty)
    }

    func testWarningIdentityIncludesTheProviderSoTwoTracksDoNotCollide() {
        let windows = [window("Claude", "five_hour", 90), window("Codex", "five_hour", 90)]
        let warnings = ProviderOverview.warnings(windows: windows, paces: [],
                                                 approaching: 60, atLimit: 85)
        XCTAssertEqual(Set(warnings.map(\.id)).count, 2)
    }
}

File: windows/tests/Redline.Core.Tests/ProviderOverviewTests.cs
// Which state a provider card resolves to, and what it says when a figure is missing.
using Redline.Core;

namespace Redline.Core.Tests;

public sealed class ProviderCardStateTests
{
    private static ProviderUsage Usage(int io, double cost = 1) => new() { Io = io, Cost = cost };

    [Fact]
    public void AMissingToolIsNotInstalledWhateverElseIsTrue()
    {
        var card = ProviderOverview.Card("Codex", installed: false, read: true, usage: Usage(5000));
        Assert.Equal(ProviderConnection.NotInstalled, card.Connection);
        // a tool that is not on this machine has no figures to draw
        Assert.False(card.Connection.HasFigures());
    }

    [Fact]
    public void AnInstalledToolSwitchedOffReadsAsNotRead()
    {
        var card = ProviderOverview.Card("Codex", installed: true, read: false, usage: Usage(5000));
        Assert.Equal(ProviderConnection.NotRead, card.Connection);
    }

    [Fact]
    public void UnreachableOutranksIdleSoAStoppedServerIsNeverReportedAsQuiet()
    {
        // The bug this guards: a stopped Ollama has no usage, and "no usage in this range"
        // reads as a quiet day rather than as a server that is not running.
        var card = ProviderOverview.Card("Ollama", installed: true, read: true, reachable: false, usage: null);
        Assert.Equal(ProviderConnection.Unreachable, card.Connection);
    }

    [Fact]
    public void NoUsageAndNoWindowIsIdle()
    {
        var card = ProviderOverview.Card("Codex", installed: true, read: true, usage: null);
        Assert.Equal(ProviderConnection.Idle, card.Connection);
    }

    [Fact]
    public void AWindowAloneMakesACardActiveEvenWithNoTokens()
    {
        // A claude.ai user has rate limits and no transcripts, and the card still has news
        var window = new LimitWindow("Claude", "five_hour", 40, DateTimeOffset.UtcNow.AddSeconds(3600));
        var card = ProviderOverview.Card("Claude", installed: true, read: true, usage: null, windows: new[] { window });
        Assert.Equal(ProviderConnection.Active, card.Connection);
        Assert.Equal(40.0, card.Utilization);
    }

    [Fact]
    public void TheWorstWindowIsTheOneShown()
    {
        var session = new LimitWindow("Codex", "five_hour", 22, null);
        var week = new LimitWindow("Codex", "seven_day", 71, null);
        var card = ProviderOverview.Card("Codex", installed: true, read: true, usage: Usage(10),
                                         windows: new[] { session, week });
        Assert.Equal("seven_day", card.WorstWindow?.Key);
        Assert.Equal(29.0, card.RemainingPercent);
    }

    [Fact]
    public void WindowsBelongingToOtherProvidersAreIgnored()
    {
        var mine = new LimitWindow("Codex", "five_hour", 10, null);
        var theirs = new LimitWindow("Claude", "five_hour", 99, null);
        var card = ProviderOverview.Card("Codex", installed: true, read: true, usage: Usage(10),
                                         windows: new[] { mine, theirs });
        Assert.Equal(10.0, card.Utilization); // another provider's window leaked into this card
    }

    [Fact]
    public void UninformativeWindowsAreDropped()
    {
        var junk = new LimitWindow("Claude", "nimbus_quill", 0, null);
        var card = ProviderOverview.Card("Claude", installed: true, read: true, usage: null, windows: new[] { junk });
        Assert.Null(card.WorstWindow);
        Assert.Equal(ProviderConnection.Idle, card.Connection);
    }

    [Fact]
    public void AStaleReadingCarriesNoPaceBecauseARateNeedsACurrentNumber()
    {
        var resets = DateTimeOffset.UtcNow.AddSeconds(3600);
        var window = new LimitWindow("Claude", "five_hour", 50, resets);
        var pace = PaceEstimator.Pace(window);
        Assert.NotNull(pace); // fixture needs a pace for the test to mean anything
        var card = ProviderOverview.Card("Claude", installed: true, read: true, usage: null,
                                         windows: new[] { window }, paces: new[] { pace! },
                                         asOf: DateTimeOffset.UtcNow.AddSeconds(-3600));
        Assert.True(card.IsStale);
        Assert.Null(card.Pace);
    }

    [Fact]
    public void AFreshReadingKeepsItsPace()
    {
        var resets = DateTimeOffset.UtcNow.AddSeconds(3600);
        var window = new LimitWindow("Claude", "five_hour", 50, resets);
        var pace = PaceEstimator.Pace(window);
        Assert.NotNull(pace); // fixture needs a pace
        var card = ProviderOverview.Card("Claude", installed: true, read: true, usage: null,
                                         windows: new[] { window }, paces: new[] { pace! },
                                         asOf: DateTimeOffset.UtcNow);
        Assert.False(card.IsStale);
        Assert.NotNull(card.Pace);
    }

    [Fact]
    public void StaleOutranksTheStatusItWouldOtherwiseDraw()
    {
        var window = new LimitWindow("Claude", "five_hour", 96, null);
        var card = ProviderOverview.Card("Claude", installed: true, read: true, usage: null,
                                         windows: new[] { window }, asOf: DateTimeOffset.UtcNow.AddSeconds(-4000));
        // an old reading must never be drawn as a live one
        Assert.Equal(RLStatusKind.Stale, card.Status(approaching: 60, atLimit: 85).Kind);
    }

    [Fact]
    public void ALocalProviderExplainsWhyItHasNoLimitRatherThanShowingNothing()
    {
        var card = ProviderOverview.Card("Ollama", installed: true, read: true, reachable: true, usage: Usage(900));
        Assert.Equal(ProviderConnection.Active, card.Connection);
        Assert.Null(card.Utilization);
        Assert.Equal("Runs on this PC, so there is no rate limit to report", card.LimitNote);
    }

    [Fact]
    public void AHostedProviderWithNoWindowSaysSoAndPrefersTheCallersReason()
    {
        var plain = ProviderOverview.Card("Claude", installed: true, read: true, usage: Usage(900));
        Assert.Equal("No limit window reported yet", plain.LimitNote);

        var given = ProviderOverview.Card("Claude", installed: true, read: true, usage: Usage(900),
                                          limitsNote: "rate limited, retrying");
        Assert.Equal("rate limited, retrying", given.LimitNote);
    }

    [Fact]
    public void ACardWithAWindowCarriesNoLimitNote()
    {
        var window = new LimitWindow("Codex", "five_hour", 12, null);
        var card = ProviderOverview.Card("Codex", installed: true, read: true, usage: null,
                                         windows: new[] { window }, limitsNote: "should not be shown");
        Assert.Null(card.LimitNote);
    }

    [Fact]
    public void AnAbsentToolSaysNothingAboutLimitsBecauseTheCardAlreadySaysWhy()
    {
        var card = ProviderOverview.Card("Codex", installed: false, read: true, usage: null, limitsNote: "irrelevant");
        Assert.Null(card.LimitNote);
    }

    [Fact]
    public void EveryConnectionStateSaysSomethingAndHasATone()
    {
        foreach (var state in Enum.GetValues<ProviderConnection>())
        {
            Assert.False(string.IsNullOrEmpty(state.Phrase()));
            _ = state.Tone();
        }
    }
}

public sealed class OverviewWarningTests
{
    private static LimitWindow Window(string provider, string key, double pct, DateTimeOffset? resets = null) =>
        new(provider, key, pct, resets);

    [Fact]
    public void AHealthyWindowRaisesNothing()
    {
        var warnings = ProviderOverview.Warnings(new[] { Window("Claude", "five_hour", 12) },
            Array.Empty<Pace>(), approaching: 60, atLimit: 85);
        Assert.Empty(warnings); // a banner that is always lit says nothing
    }

    [Fact]
    public void CrossingTheConfiguredThresholdRaisesOne()
    {
        var warnings = ProviderOverview.Warnings(new[] { Window("Claude", "five_hour", 62) },
            Array.Empty<Pace>(), approaching: 60, atLimit: 85);
        Assert.Single(warnings);
        Assert.Equal(OverviewWarningKind.Approaching, warnings[0].Kind);
        Assert.Contains("62%", warnings[0].Text);
    }

    [Fact]
    public void TheThresholdsAreTheConfiguredOnesNotHardcodedSixtyAndEightyFive()
    {
        var warnings = ProviderOverview.Warnings(new[] { Window("Claude", "five_hour", 45) },
            Array.Empty<Pace>(), approaching: 40, atLimit: 70);
        Assert.Equal(OverviewWarningKind.Approaching, warnings.FirstOrDefault()?.Kind);

        var atLimit = ProviderOverview.Warnings(new[] { Window("Claude", "five_hour", 72) },
            Array.Empty<Pace>(), approaching: 40, atLimit: 70);
        Assert.Equal(OverviewWarningKind.AtLimit, atLimit.FirstOrDefault()?.Kind);
    }

    [Fact]
    public void WorstNewsIsReadFirst()
    {
        var windows = new[] { Window("Claude", "five_hour", 65), Window("Codex", "seven_day", 91) };
        var warnings = ProviderOverview.Warnings(windows, Array.Empty<Pace>(), approaching: 60, atLimit: 85);
        Assert.Equal(OverviewWarningKind.AtLimit, warnings[0].Kind);
        Assert.Equal("Codex", warnings[0].Provider);
    }

    [Fact]
    public void AStaleProviderRaisesNothingBecauseItIsNotEvidenceAboutNow()
    {
        var warnings = ProviderOverview.Warnings(new[] { Window("Claude", "five_hour", 99) },
            Array.Empty<Pace>(), approaching: 60, atLimit: 85, staleProviders: new HashSet<string> { "Claude" });
        Assert.Empty(warnings);
    }

    [Fact]
    public void AWindowRunningOutBeforeItResetsIsWorthSayingEvenWhileHealthy()
    {
        var resets = DateTimeOffset.UtcNow.AddSeconds(4 * 3600);
        var w = Window("Codex", "five_hour", 20, resets);
        // Burning fast enough to hit the cap well before the reset
        var pace = new Pace("Codex", "five_hour", 20, resets, 80, new PaceBasis.WindowAverage(), 0.2,
                            DateTimeOffset.UtcNow.AddSeconds(3600));
        var warnings = ProviderOverview.Warnings(new[] { w }, new[] { pace }, approaching: 60, atLimit: 85);
        Assert.Single(warnings);
        Assert.Equal(OverviewWarningKind.RunsOutEarly, warnings[0].Kind);
        Assert.Contains("before it resets", warnings[0].Text);
    }

    [Fact]
    public void UninformativeWindowsRaiseNothing()
    {
        var junk = Window("Claude", "nimbus_quill", 0);
        Assert.Empty(ProviderOverview.Warnings(new[] { junk }, Array.Empty<Pace>(), approaching: 60, atLimit: 85));
    }

    [Fact]
    public void WarningIdentityIncludesTheProviderSoTwoTracksDoNotCollide()
    {
        var windows = new[] { Window("Claude", "five_hour", 90), Window("Codex", "five_hour", 90) };
        var warnings = ProviderOverview.Warnings(windows, Array.Empty<Pace>(), approaching: 60, atLimit: 85);
        Assert.Equal(2, warnings.Select(w => w.Id).ToHashSet().Count);
    }
}
